File: src/bot/commands/clan/boosts.ts
import { AutocompleteInteraction, SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { EMOJIS } from '../../../util/emojis.js';

/** Shows Super Troop boosts active among a clan's members. */
export default class BoostsCommand extends Command {
  public constructor() {
    super('boosts', { category: 'clan', description: 'Show active Super Troop boosts in a clan.', defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('boosts')
      .setDescription('Show active Super Troop boosts in a clan.')
      .addStringOption((opt) => opt.setName('tag').setDescription('Clan tag or alias.').setAutocomplete(true));
  }

  public async autocomplete(interaction: AutocompleteInteraction, args: Record<string, any>) {
    await interaction.respond(await this.client.resolver.clanAutocomplete(interaction.guildId, String(args.tag ?? '')));
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: clan, error } = await this.client.resolver.resolveClan(interaction, args.tag);
    if (!clan) return interaction.editReply(error!);

    const players = await Promise.all(clan.members.map((m) => this.client.coc.getPlayer(m.tag)));
    const boostMap = new Map<string, string[]>();
    for (const p of players) {
      if (!p) continue;
      const active = p.troops.filter((t) => t.isSuperTroop && t.isActive);
      for (const t of active) {
        const list = boostMap.get(t.name) ?? [];
        list.push(p.name);
        boostMap.set(t.name, list);
      }
    }

    const embed = baseEmbed(this.client, interaction).setTitle(`${clan.name} (${clan.tag}) — Active Boosts`);
    if (!boostMap.size) {
      embed.setDescription('No active Super Troop boosts right now.');
    } else {
      embed.setDescription(
        [...boostMap.entries()]
          .map(([troop, names]) => `${EMOJIS.TROOP} **${troop}** — ${names.length}\n${names.map((n) => `• ${n}`).join('\n')}`)
          .join('\n\n')
          .slice(0, 4000)
      );
    }

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/clan/commands/boosts.py
"""/clan boosts, active Super Troop boosts among clan members."""

import asyncio

import discord

from utils.embeds import base_embed, error_embed
from utils.emojis import E_TROOP
from utils.resolver import resolve_clan
from utils.units import active_super_troops


async def boosts(interaction: discord.Interaction, tag: str = None):
    await interaction.response.defer()
    try:
        clan = await resolve_clan(interaction, tag)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    coc_client = interaction.client.coc_client

    async def fetch(member_tag):
        try:
            return await coc_client.get_player(member_tag)
        except Exception:
            return None

    players = await asyncio.gather(*(fetch(m.tag) for m in clan.members))

    boost_map: dict[str, list[str]] = {}
    for p in players:
        if not p:
            continue
        for troop in active_super_troops(p):
            boost_map.setdefault(troop["name"], []).append(p.name)

    embed = await base_embed(interaction, title=f"{clan.name} ({clan.tag}), Active Boosts")
    if not boost_map:
        embed.description = "No active Super Troop boosts right now."
    else:
        blocks = []
        for troop_name, names in boost_map.items():
            listing = "\n".join(f"• {n}" for n in names)
            blocks.append(f"{E_TROOP} **{troop_name}**, {len(names)}\n{listing}")
        embed.description = "\n\n".join(blocks)[:4000]

    await interaction.followup.send(embed=embed)
